### PR TITLE
Manage NodePortProxy using Kubermatic Operator

### PR DESCRIPTION
**Pull request by xrstf**
**What this PR does / why we need it**:
Kubermatic needs the nodeport-proxy (NPP) in each seed to function properly, and the NPP is useless without Kubermatic. The operator is already reconciling each Seed, so it would be nice if it also quickly installed the NPP. If there's a 1:1 relation between the two, why not merge them?

This PR makes it so that the operator also installs the NPP. It will then live inside the Kubermatic namespace, so it's explicitly not compatible with the old Helm chart. This means that it will not 1:1 replace it or try to migrate. The intention is that new customers use the new NPP and existing customers should slowly migrate while the NPP chart is deprecated. Running both NPP installations in parallel is possible (you end up with 2 identical LoadBalancer services), so migrations can happen slowly over time. If customers really do not want the new NPP, they can always disable it per seed in the Seed CRD.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: api/pkg/controller/operator/common/defaults.go [hunk 1]
```diff
@@ -481,6 +525,44 @@ func DefaultConfiguration(config *operatorv1alpha1.KubermaticConfiguration, logg
 	return copy, nil
 }
 
+func DefaultSeed(seed *kubermaticv1.Seed, logger *zap.SugaredLogger) (*kubermaticv1.Seed, error) {
+	logger = logger.With("seed", seed.Name)
+	logger.Debug("Applying defaults to Seed")
+
+	copy := seed.DeepCopy()
```
Comment (moadqassem) [inline api/pkg/controller/operator/common/defaults.go:532]: Nit: Isn't this a builtin function? maybe a better naming?

File: api/pkg/controller/operator/common/defaults.go [hunk 2]
```diff
@@ -120,6 +123,47 @@ var (
...
+	}
+
+	DefaultNodeportProxyUpdaterResources = corev1.ResourceRequirements{
+		Requests: corev1.ResourceList{
+			corev1.ResourceCPU:    resource.MustParse("50m"),
+			corev1.ResourceMemory: resource.MustParse("32Mi"),
+		},
+		Limits: corev1.ResourceList{
+			corev1.ResourceCPU:    resource.MustParse("150m"),
+			corev1.ResourceMemory: resource.MustParse("32Mi"),
```
Comment (moadqassem) [inline api/pkg/controller/operator/common/defaults.go:155]: Nit: maybe `48Mi` to be different than what it requests?

File: api/pkg/controller/operator/seed/resources/nodeportproxy/rbac.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+)
+
+const (
+	RoleName        = "nodeport-proxy"
+	RoleBindingName = "nodeport-proxy"
+)
+
+func RoleCreator(cfg *operatorv1alpha1.KubermaticConfiguration) reconciling.NamedRoleCreatorGetter {
```
Comment (moadqassem) [inline api/pkg/controller/operator/seed/resources/nodeportproxy/rbac.go]: `cfg` is an unused parameter ;)

File: api/pkg/controller/operator/seed/resources/nodeportproxy/rbac.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+					Name:      ServiceAccountName,
+				},
+			}
+
+			return crb, nil
+		}
+	}
+}
+
+func ServiceAccountCreator(cfg *operatorv1alpha1.KubermaticConfiguration) reconciling.NamedServiceAccountCreatorGetter {
```
Comment (moadqassem) [inline api/pkg/controller/operator/seed/resources/nodeportproxy/rbac.go:103]: same goes here `cfg` is an unused parameter ;)

File: api/pkg/controller/operator/seed/resources/nodeportproxy/deployments.go [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+)
+
+const (
+	ServiceAccountName    = "nodeport-proxy"
+	EnvoyDeploymentName   = "nodeport-proxy-envoy"
+	UpdaterDeploymentName = "nodeport-proxy-updater"
+	EnvoyPort             = 8002
+)
+
+func EnvoyDeploymentCreator(cfg *operatorv1alpha1.KubermaticConfiguration, seed *kubermaticv1.Seed, versions common.Versions) reconciling.NamedDeploymentCreatorGetter {
```
Comment (moadqassem) [inline api/pkg/controller/operator/seed/resources/nodeportproxy/deployments.go]: `cfg` parameter is not used.

File: api/pkg/controller/operator/seed/resources/nodeportproxy/deployments.go [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+				MatchLabels: map[string]string{
+					common.NameLabel: EnvoyDeploymentName,
+				},
+			}
+			return pdb, nil
+		}
+	}
+}
+
+func UpdaterDeploymentCreator(cfg *operatorv1alpha1.KubermaticConfiguration, seed *kubermaticv1.Seed, versions common.Versions) reconciling.NamedDeploymentCreatorGetter {
```
Comment (moadqassem) [inline api/pkg/controller/operator/seed/resources/nodeportproxy/deployments.go]: `cfg` parameter is not used.

File: api/pkg/controller/operator/common/defaults.go (hunk 2 above)
Comment (xrstf) [reply api/pkg/controller/operator/common/defaults.go:155]: I don't understand that reasoning at all :-D Use a different value just to be different? What's bad about a fixed-size (non-burstable) pod?

Comment (moadqassem) [reply api/pkg/controller/operator/common/defaults.go:155]: Sorry 😂! my expressing ability has failed me( happening quite a lot lately 🤔). I thought that you have actually forgotten to adjust this as the other resource( the `DefaultNodeportProxyEnvoyManagerResources`).

File: api/pkg/controller/operator/common/defaults.go (hunk 1 above)
Comment (xrstf) [reply api/pkg/controller/operator/common/defaults.go:532]: What do you mean? Yes, the DeepCopy() is auto-generated.

Comment (moadqassem) [reply api/pkg/controller/operator/common/defaults.go:532]: `copy` is a builtin function

**Comment by xrstf** [comment]
PR feedback integrated.

**Pull request merged**